### PR TITLE
[WIP] Feature/split login strategies

### PR DESCRIPTION
**Pull request by ClemensSahs**
Here is a first idea for splitting the auth strategies, feedback are welcome ;)

future steps:
- [x] enable and disable exists strategies (like facebook yes, google no, github yes)
- [x] allow custom overwrite (auto detect or passport strategies register)
- [x] connect to the view (we don't render disabled strategies)
- [ ] write test
- feedback for more ideas wanted.

add more strategies (logic)
- [x] facebook
- [x] github
- [ ] google
- [ ] twitter

best regards

### CODE REVIEW

**Comment by advance512** [comment]
I would suggest you introduce a directory hierarchy, and split the various strategies to files.

**Comment by ClemensSahs** [comment]
@advance512
yes split in to separate files is in my mind a good solution. Subdirectories for each methods seems like a overkill.
Hope I have found a good solution for all use cases.

**Comment by ClemensSahs** [comment]
@koistya I'm ready for review

File: src/config.js [hunk 1]
```diff
@@ -26,6 +26,54 @@ export const analytics = {
...
+    usedAuth: process.env.AUTH_CONFIG_USED_AUTH || 'facebook,twitter,google,github',
+    loginButtons: {
+      facebook: {
+        icon: 'M22 16l1-5h-5V7c0-1.544.784-2 3-2h2V0h-4c-4.072 0-7 2.435-7 7v4H7v5h5v14h6V16h4z',
+        routeTo: 'login/facebook',
+        buttonClass: 'facebook',
+        buttonText: 'Log in with Facebook',
+      },
+      google: {
+        icon: 'M30 13h-4V9h-2v4h-4v2h4v4h2v-4h4m-15 2s-2-1.15-2-2c0 0-.5-1.828 1-3 ' +
```
Comment (langpavel) [inline src/config.js]: I'm really sure that this should be separate svg file

File: src/config.js [hunk 2]
```diff
@@ -26,6 +26,54 @@ export const analytics = {
...
+              '0 9.97-3 9.97-7 0-2.69-2.545-4.264-5-6zm-4-4c-2.395 0-5.587-2.857-6-6C4.587 ' +
+              '3.856 6.607.93 9 1c2.394.07 4.603 2.908 5.017 6.052C14.43 10.195 13 13 11 ' +
+              '13zm-1 15c-3.566 0-7-1.29-7-4 0-2.658 3.434-5.038 7-5 .832.01 2 0 2 0 1 0 ' +
+              '2.88.88 4 2 1 1 1 2.674 1 3 0 3-1.986 4-7 4z',
+        routeTo: 'login/google',
+        buttonClass: 'google',
+        buttonText: 'Log in with google',
+      },
+      twitter: {
+        icon: 'M30 6.708c-1.105.49-2.756 1.143-4 1.292 1.273-.762 2.54-2.56 ' +
```
Comment (langpavel) [inline src/config.js]: separate svg resource...

File: src/config.js [hunk 3]
```diff
@@ -26,6 +26,54 @@ export const analytics = {
...
+              '3 0 2.136 1.348 3.894 3 5-1.01-.033-2.17-.542-3-1 0 2.98 4.186 6.432 7 7-1 ' +
+              '1-4.623.074-5 0 .784 2.447 3.31 3.95 6 4-2.105 1.648-4.647 2.51-7.53 2.51-.5 ' +
+              '0-.988-.03-1.47-.084C2.723 27.17 6.523 28 10 28c11.322 0 17-8.867 17-17 ' +
+              '0-.268.008-.736 0-1 1.2-.868 2.172-2.058 3-3.292z',
+        routeTo: 'login/twitter',
+        buttonClass: 'twitter',
+        buttonText: 'Log in with twitter',
+      },
+      github: {
+        icon: 'M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 ' +
```
Comment (langpavel) [inline src/config.js]: and this too...

File: src/config.js [hunk 4]
```diff
@@ -45,4 +93,10 @@ export const auth = {
     secret: process.env.TWITTER_CONSUMER_SECRET || 'KTZ6cxoKnEakQCeSpZlaUCJWGAlTEBJj0y2EMkUBujA7zWSvaQ',
   },
 
+  // https://github.com/
+  github: {
+    id: process.env.GITHUB_CONSUMER_KEY || 'xxxxxxxxxxxxxxxxxxxxx',
+    secret: process.env.GITHUB_CONSUMER_SECRET || 'xyxyxyxyxyxyxyxyxyxyxyxyxyxyxyxyxyxyxyxyxy',
```
Comment (langpavel) [inline src/config.js]: Very nice default strings :smile: — of course, not really. Something shorter which can be checked on runtime.. or omit default entirely can be nicer

File: src/core/passport/index.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import passport from 'passport';
+import expressJwt from 'express-jwt';
+import { Router as ExpressRouter } from 'express';
+import { auth } from '../../config';
+
+import * as passportFacebook from './facebook';
+import * as passportGoogle from './google';
+import * as passportGithub from './github';
+import * as passportTwitter from './twitter';
+import * as passportCustom from '../../custom/passport';
```
Comment (langpavel) [inline src/core/passport/index.js]: Can this be explicit?

File: src/data/schema.js [hunk 6]
```diff
@@ -14,13 +14,15 @@ import {
...
 import news from './queries/news';
+import auth from './queries/auth';
 
 const schema = new Schema({
   query: new ObjectType({
     name: 'Query',
     fields: {
       me,
       news,
+      auth,
```
Comment (langpavel) [inline src/data/schema.js]: Is this really needed? I think, can you easily implement near-infinity trusted possibilities of authentication strategies without touching code? I think, sending localization messages like I did is outer-limit, but sending authentication methods? This is really silly :laughing:

File: src/data/types/AuthType.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  fields: {
+    loginName: { type: new NonNull(StringType) },
+    icon: { type: new NonNull(StringType) },
+    buttonClass: { type: new NonNull(StringType) },
+    buttonText: { type: new NonNull(StringType) },
+    routeTo: { type: new NonNull(StringType) },
+  },
+});
+
+export default AuthType;
```
Comment (langpavel) [inline src/data/types/AuthType.js]: Most customizable CMS on the world :smile:

**Comment by ulani** [comment]
@ClemensSahs  thank you very much for this PR! Unfortunately, we have close it due to inactivity. Feel free to re-open it or [join](https://discord.com/invite/2nKEnKq) our [Discord](https://discord.com/invite/2nKEnKq) channel for [discussion](https://github.com/kriasoft/react-starter-kit/discussions/1950).

NOTE: The `main` branch has been updated with React Starter Kit v2, using JAM-style architecture.

**Pull request closed without merging**